### ISSUE TITLE
oci_containerengine_node_pool nodes.public_ip unavailable

### ISSUE THREAD

**Issue by cameronsenese** [oracle/terraform-provider-oci#582]
### Terraform Version
```
$ terraform -v
Terraform v0.11.8
```

### OCI Provider Version
```
$ \<path-to-plugin\>/terraform-provider-oci
 terraform-provider-oci 2.2.1
```

### Description:
When trying to obtain 'public_ip' attribute for oci_containerengine_node_pool worker nodes - on 'apply' operation terraform generates the following message, and fails to run:  
`* output.nod-wkr-01.ip: Resource 'oci_containerengine_node_pool.test_node_pool' does not have attribute 'nodes.public_ip' for variable 'oci_containerengine_node_pool.test_node_pool.nodes.public_ip'`  

I have input some alternate configurations for the output, in an attempt to retrieve the ip address for a specific node ( e.g. 'oci_containerengine_node_pool.test_node_pool.nodes.0.public_ip'), but receive a similar output form terraform:  
`* output.nod-wkr-01.ip: Resource 'oci_containerengine_node_pool.test_node_pool' does not have attribute 'nodes.0.public_ip' for variable 'oci_containerengine_node_pool.test_node_pool.nodes.0.public_ip'`  

I have also attempted to specify only the nodepool (e.g. 'oci_containerengine_node_pool.test_node_pool.nodes`) - in this instance the apply operation completes, but the output attribute is empty:
```
oci_containerengine_node_pool.test_node_pool: Creation complete after 4s (ID: ocid1.nodepool.oc1.iad.aaaaaaaaae4wimle...zrgmzdezbrhayweylggu4dqolbmnqwkodfmqzw)

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

Outputs:
nod-wkr-01.ip = []
```

### Terraform Plan

```
module "network" {
  source = "./module/network"
  ### Network :: VCN ###
  g001_tenancyOcid = "${var.g001_tenancyOcid}"
  g002_compartmentOcid = "${var.g002_compartmentOcid}"
  n000_vcnCidrBlock = "${var.n000_vcnCidrBlock}"
  n001_vcnDisplayName = "${var.e000_envName}${var.e001_envNumber}-vcn"
  n002_vcnDnsLabel = "${var.e000_envName}${var.e001_envNumber}"
  ### Network :: Internet Gateway ###
  n005_igwDisplayName = "${var.e000_envName}${var.e001_envNumber}-igw"
  n006_igwEnabled = "${var.n006_igwEnabled}"
  ### Network :: Route Table ###
  n010_rtbDisplayName = "${var.e000_envName}${var.e001_envNumber}-rtb"
  n011_rtbCidrBlock = "${var.n011_rtbCidrBlock}"
  ### Network :: Subnet ###
  n015_netNewbits = "${var.n015_netNewbits}"
  n016_netSubnets = "${var.n016_netSubnets}"
  ### Network :: Security List ###
  n020_secDisplayNameLbr = "${var.e000_envName}${var.e001_envNumber}-sec-lbr"
  n021_secDisplayNameWkr = "${var.e000_envName}${var.e001_envNumber}-sec-wkr"
}

resource "oci_containerengine_cluster" "test_cluster" {
	#Required
	compartment_id = "${var.g002_compartmentOcid}"
	kubernetes_version = "${var.k009_kubeVers}"
	name = "${var.k000_clusterName}"
	vcn_id = "${module.network.vcn-01.id}"
	#Optional
	options {
		add_ons {
			is_kubernetes_dashboard_enabled = true
			is_tiller_enabled = true
		}
		service_lb_subnet_ids = ["${module.network.net-lbr-01.id}", "${module.network.net-lbr-02.id}"]
	}
}

resource "oci_containerengine_node_pool" "test_node_pool" {
	#Required
	cluster_id = "${oci_containerengine_cluster.test_cluster.id}"
	compartment_id = "${var.g002_compartmentOcid}"
	kubernetes_version = "${var.k009_kubeVers}"
	name = "${var.k000_clusterName}"
	node_image_name = "${var.k008_wkrNodeImage}"
	node_shape = "${var.k007_wkrNodeShape}"
	subnet_ids = ["${module.network.net-wkr-01.id}"]
	#subnet_ids = ["${module.network.net-wkr-01.id}", "${module.network.net-wkr-02.id}", "${module.network.net-wkr-03.id}"]
	quantity_per_subnet = "${var.k002_quantityPerSubnet}"
	ssh_public_key = "${file(var.c004_rsaPublicKeyPath)}"
}

output "nod-wkr-01.ip" {
  value= "${oci_containerengine_node_pool.test_node_pool.nodes.public_ip}"
}
```

**Comment by kilokahn** [comment]
@cameronsenese thanks for patiently waiting on this. The additional information you provided is quite helpful and from the looks of it, it would seem that the `node_pool.nodes` being empty is the problem. Can you also provide the debug logs information for the invocation? You can do it by
`DEBUG=true TF_LOG=DEBUG OCI_GO_SDK_DEBUG=1 terraform apply &> /tmp/node_pools_empty.log`

**Comment by kilokahn** [comment]
@cameronsenese - i unaware of the investigation that @ravitandonrt and @parweza had done for this. It seems that the service is returning the lifecycle state as being successful immediately for the node pool, but the nodes in the node pool are still coming up. Unless the service API fixes this gap, we don't have a good solution. We are following up with the service team on this.

**Comment by cameronsenese** [comment]
Hi @kilokahn, Thank you for the confirmation.
If you are interested, I have written a solution in Golang that addresses this gap, named `oke-go`

oke-go is a CLI utility designed for use with Oracle Container Engine for Kubernetes (OKE).  
oke-go provides a command-line interface for interaction with OKE, including Kubernetes cluster lifecycle.  

oke-go utilises the OCI Go SDK to interact with OKE.  
oke-go provides a flag for it's `createOkeCluster` command named `--waitNodesActive`.  
This flag provides a mechanism to request node pool creation, then wait until nodes in the node pool are active before returning. _It actually supports waiting specifically for any nodes to be active, all nodes to be active, or to not wait at all.._

I have picked out some logic and a function from oke-go that demonstrates the mechanism:

``` go
// wait for create node completion - all nodes..
if *g3WaitNodesActive != "false" {
	nodeProvisioning := true
	// This loop continues while "nodeProvisioning" is true..
	for nodeProvisioning {
		getNodeLifeCycleState(ctx, c, *g3NodePoolId)
		// if we are waiting for all nodes to be active..
		if *g3WaitNodesActive == "all" {
			boolLcs := (strings.Contains(nodeLifeCycleState, "ATING"))
				if nodeLifeCycleState != "{}" {
					if boolLcs != true {
						nodeProvisioning = false
					}
				}
		}
		// if we are waiting for any nodes to be active..
		if *g3WaitNodesActive == "any" {
			boolLcs := (strings.Contains(nodeLifeCycleState, "ACTIVE"))
				if nodeLifeCycleState != "{}" {
					if boolLcs == true {
						nodeProvisioning = false
					}
				}
		}
			time.Sleep(15 * time.Second)
	}
}
```

``` go
// get worker node lifecycle status..
func getNodeLifeCycleState(
	ctx context.Context,
	client containerengine.ContainerEngineClient,
	nodePoolId string) containerengine.GetNodePoolResponse {

	req := containerengine.GetNodePoolRequest{}
	req.NodePoolId = common.String(nodePoolId)

	resp, err := client.GetNodePool(ctx, req)
	helpers.FatalIfError(err)

	// marshal & parse json..
	nodePoolResp := resp.NodePool
	nodesJson, _ := json.Marshal(nodePoolResp)
	jsonParsed, _ := gabs.ParseJSON(nodesJson)
	nodeLifeCycleState = (jsonParsed.Path("nodes.lifecycleState").String())

	return resp
}
```

Do let me know if you are interested - I can point you to the souce for the `oke-go` utility.

Rgds..

**Comment by kilokahn** [comment]
Thanks for sharing the snips from the utility. While we could do something like this, it makes sense for the service to support it from upstream, and hence it might not be in the best interest to put this in the provider code. That said, @ravitandonrt is following up on this.

**Comment by cameronsenese** [comment]
Cheers @kilokahn, agree that upstream support would be an ideal solution.

**Comment by PraveenViswanathan** [comment]
Hello all,

Do we have any progress on this issue. Here' a similar one.

I am creating a node pool with one worker node and requirement is to get the worker node details from the nodes argument. As I know even if oci_containerengine_node_pool resource is created, worker nodes might not have created so I am using local-exec to wait for some time and try to get the worker node details using below code in my output.tf

"${oci_containerengine_node_pool.np.*.nodes[0].public_ip}"

But I am still ending up with below message during the apply phase. 

-------------------------------------------------------------------------------------
Apply Error: Invalid index
  on modules/cluster/output.tf line 2, in output "public-ip1" 
   2:     value= "${oci_containerengine_node_pool.webapp-np.*.nodes.0.public_ip}"
The given key does not identify an element in this collection value.
-------------------------------------------------------------------------------------

When I checked the terraform state the node pool resource is showing nodes as empy list ([]). 

Why when even the worker nodes are available terraform state file is not showing it ?
What is the way to get the worked nodes created by a oci_containerengine_node_pool.

**Comment by ravinitp** [comment]
We are very sorry that we couldn't respond to each and every issue reported on GitHub. 
Although we have refined the process to prioritize customer issues on GitHub, since this issue was reported a while ago, there is a good chance it may have been fixed in the latest version of Terraform Provider OCI. 
If you are still experiencing this issue, please create a new issue and label it as Bug.
